### ISSUE TITLE
Files.newOutputStream(<S3Path>) fails with a java.lang.UnsupportedOperationException.

### ISSUE THREAD

**Issue by sielenk-yara** [awslabs/aws-java-nio-spi-for-s3#158]
### Context

Version used:

    implementation("software.amazon.nio.s3:aws-java-nio-spi-for-s3:1.2.3")

on JDK `liberica-11.0.18-x86`.

### Problem

The call goes through `java.nio.file.Files#newOutputStream` into `java.nio.file.spi.FileSystemProvider#newOutputStream` where the empty option array is detected and `java.nio.file.spi.FileSystemProvider#DEFAULT_OPEN_OPTIONS` (an _immutable_ set) is passed into the call to `newByteChannel` and from there on to ultimately `S3WritableByteChannel#S3WritableByteChannel(S3Path, S3AsyncClient, Set<? extends OpenOption>, Long, TimeUnit)`.

In there is the unconditional attempt to remove one value

      [...]
      options.remove(StandardOpenOption.CREATE_NEW);
      [...]
      
which fails for the immutable set (even though the value is not in there and no actual change would happen).

### Workaround

Pass the default options explicitly:

    [...]
    Files.newOutputStream(<path>,
        StandardOpenOption.CREATE,
        StandardOpenOption.TRUNCATE_EXISTING,
        StandardOpenOption.WRITE
    )
    [...]

**Comment by sielenk-yara** [comment]
The code is here: https://github.com/awslabs/aws-java-nio-spi-for-s3/blob/98e2f9da844b12443b1841373a0643a4e56c576a/src/main/java/software/amazon/nio/spi/s3/S3WritableByteChannel.java#L77

**Comment by onetom** [comment]
I just hit this issue too yesterday too.

I was searching the issues for `ImmutableCollections` &  `uoe` and couldn't find this issue, so here is a stacktrace for better discoverability:
```
Execution error (UnsupportedOperationException) at java.util.ImmutableCollections/uoe (ImmutableCollections.java:142).
null
java.lang.UnsupportedOperationException
	at java.base/java.util.ImmutableCollections.uoe(ImmutableCollections.java:142)
	at java.base/java.util.ImmutableCollections$AbstractImmutableCollection.remove(ImmutableCollections.java:150)
	at software.amazon.nio.spi.s3.S3WritableByteChannel.<init>(S3WritableByteChannel.java:77)
	at software.amazon.nio.spi.s3.S3SeekableByteChannel.<init>(S3SeekableByteChannel.java:105)
	at software.amazon.nio.spi.s3.S3SeekableByteChannel.<init>(S3SeekableByteChannel.java:78)
	at software.amazon.nio.spi.s3.S3FileSystemProvider.newByteChannel(S3FileSystemProvider.java:241)
	at software.amazon.nio.spi.s3.S3FileSystemProvider.newByteChannel(S3FileSystemProvider.java:216)
	at java.base/java.nio.file.spi.FileSystemProvider.newOutputStream(FileSystemProvider.java:484)
	at java.base/java.nio.file.Files.newOutputStream(Files.java:228)
	at java.base/java.nio.file.Files.write(Files.java:3511)
	at babashka.fs$write_bytes.invokeStatic(fs.cljc:1079)
	at babashka.fs$write_bytes.invoke(fs.cljc:1056)
	at babashka.fs$write_bytes.invokeStatic(fs.cljc:1071)
	at babashka.fs$write_bytes.invoke(fs.cljc:1056)
...
```

I would also argue, that the `.newOutputStream` logic in the JDK is a bit questionable, since it either passes an immutable `java.util.Set` or a mutable `java.util.HashSet` to the `abstract newByteChannel`, which expects a `Set`, implying immutability:
https://github.com/openjdk/jdk/blob/e3c6cf8eaf931d9eb46b429a5ba8d3bbded3728a/src/java.base/share/classes/java/nio/file/spi/FileSystemProvider.java#L472

I guess it's whole `Set` business is just complected in general...
`Set` is an interface to something *mutable*, because its methods, like `add` and `remove` doesn't return new `Set` objects, but then it's _also_ a container for some factory functions, like `of` and `copyOf`, which return *immutable* objects.

**Comment by markjschreiber** [comment]
I think we can add a simple guard here

```java
if(!options.isEmpty()) {
      // testing for empty removes the chance that we will try to remove the option from EMPTY_MAP
      // which is immutable.
      options.remove(StandardOpenOption.CREATE_NEW);
}
```

**Comment by markjschreiber** [comment]
Unfortunately it seems there are other ways that `options.remove()` will fail, such as unmodifiable views. So a try catch is going to be needed.

**Comment by markjschreiber** [comment]
Should be fixed with PR 246. Please re-open if this is still a problem.

**Comment by sielenk-yara** [comment]
That fix looks to me like on one hand overkill and on the other hand not _really_ a fix.
Passing an immutable `Set` still can lead to an exception but there is no _real_ reason to require it to be mutable.

What about something like this:

      channel = Files.newByteChannel(
          this.tempFile,
          options.contains(StandardOpenOption.CREATE_NEW)
              ? options.stream().filter(StandardOpenOption.CREATE_NEW::equals).collect(Collectors.toSet())
              : options);

It might even be a good idea to _always_ copy the set to avoid nasty object reuse bugs...

**Comment by sielenk-yara** [comment]
I think I am not allowed to reopen this, @markjschreiber.

**Comment by sielenk-yara** [comment]
It indeed occurred to me that there was a negation missing after I left yesterday. I had a correct lambda there before (and even a unit test!) and thought I was smart and it was late... 🤷‍♀️

The creation of an immutable set is fine by me. I'd even say it should be the default!

PR best on top of main, I guess? I did my changes in the 1.2.3 code.

**Comment by sielenk-yara** [comment]
My bad ... did not look at the latest code, only at the first change in the history after my initial investigation.
This has already been fixed nicely by @guicamest (#296).

Sorry for the inconvenience!
